packages/databricks-vscode/src/bundle/BundleProjectManager.ts

```diff
@@ -0,0 +1,254 @@
+    private logger = logging.NamedLogger.getOrCreate(Loggers.Extension);
+    private disposables: Disposable[] = [];
+
+    private _isBundleProject = new CachedValue<boolean>(async () => {
+        const rootBundleFile = await this.bundleFileSet.getRootFile();
+        return rootBundleFile !== undefined;
+    });
+
+    public onDidChangeStatus = this._isBundleProject.onDidChange;
+
+    private _isLegacyProject = new CachedValue<boolean>(async () => {
+        // TODO
+        return false;
+    });
+
+    private _subProjects = new CachedValue<{absolute: Uri; relative: Uri}[]>(
+        async () => {
+            const projects = await this.bundleFileSet.getSubProjects();
+            this.logger.debug(
+                `Detected ${projects.length} sub folders with bundle projects`
+            );
+            this.customWhenContext.setSubProjectsAvailable(projects.length > 0);
+            return projects;
+        }
+    );
+
+    private projectServicesReady = false;
+    private projectServicesMutex = new Mutex();
+
+    constructor(
+        private cli: CliWrapper,
+        private customWhenContext: CustomWhenContext,
+        private connectionManager: ConnectionManager,
+        private configModel: ConfigModel,
+        private bundleFileSet: BundleFileSet,
+        private workspaceUri: Uri
+    ) {
+        this.disposables.push(
+            this.bundleFileSet.bundleDataCache.onDidChange(async () => {
+                try {
+                    await this._isBundleProject.refresh();
+                } catch (error) {
+                    this.logger.error(
+                        "Failed to refresh isBundleProject var",
+                        error
+                    );
+                }
+            }),
+            this._isBundleProject.onDidChange(async () => {
+                try {
+                    await this.configureBundleProject();
+                } catch (error) {
+                    this.logger.error(
+                        "Failed to configure bundle project after isBundleProject change",
+                        error
+                    );
+                }
+            })
+        );
+    }
+
+    public async isBundleProject(): Promise<boolean> {
+        return await this._isBundleProject.value;
+    }
+
+    public async configureWorkspace(): Promise<void> {
+        // We listen to _isBundleProject changes and call configureBundleProject
+        await this._isBundleProject.refresh();
+
+        // The cached value updates subProjectsAvailabe context.
+        // We have a configurationView that shows "open project" button if the context value is true.
+        await this._subProjects.refresh();
+
+        const isLegacyProject = await this._isLegacyProject.value;
+        if (isLegacyProject) {
+            this.logger.debug(
+                "Detected a legacy project.json, starting automatic migration"
+            );
+            await this.migrateProjectJsonToBundle();
+        }
```

Comment:
What if it is both a legacy project and a bundleProject at the same time? Can both flows trigger at the same time for that case?

Reply:
Maybe inverting the order of operations here helps. We return if it is a legacy project and then refresh other caches. 

Ideally we should be able to not depend on refreshes for critical flows.

Reply:
It depends on how we want to do the migration. It might be safer not to remove existing project.json even after we created bundle config based on it. In that case the project is both "legacy" and "bundle".

I think here it makes sense to return after the `this._isBundleProject.refresh()` if we are in a bundle project.

Reply:
We might not want to remove the project.json for sometime, in case users want to revert to older extension.